### PR TITLE
12259 concat ws

### PR DESCRIPTION
**Pull request by rc-enfuse**
## Summary of the changes / Why this improves CrateDB
Added concat_ws function to the scalar functions library. User is able to concatenate with a delimiter expanding crateDB's features. https://github.com/crate/crate/issues/12259

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by amotl** [comment]
ok to test

**Comment by rc-enfuse** [comment]
CrateDB team,
I ran the following command locally against this branch and it passes without error. 
./gradlew :server:test -Dtests.seed=2DFFC6BE4D571E5 -Dtests.class=io.crate.integrationtests.disruption.replication.logical.SubscriptionDisruptionIT -Dtests.method="test_subscription_metadata_tracker_retries_on_publisher_disconnect" -Dtests.locale=se-SE -Dtests.timezone=Indian/Chagos
Not sure where the differences would be between our environments. Any help would be greatly appreciated. Thanks

**Comment by BaurzhanSakhariev** [comment]
> I ran the following command locally against this branch and it passes without error.

It seems to be a flaky test, not related to your PR, you can safely ignore it.

File: server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java:108]: I think we can avoid `endsWith/substring` calls for every row in the following way (roughly):
```
boolean firstNonNullProcessed = false;
for (...) {
    if (val != null) {
        if (!firstNonNullProcessed) {
            firstNonNullProcessed = true;
            sb.append(value);
        } else {
             sb.append(separator).append(value).
        }
    }    
}
```

File: server/src/test/java/io/crate/expression/scalar/ConcatwsFunctionTest.java
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/ConcatwsFunctionTest.java:45]: assuming that we would remove special handling of the last element, `testManyStrings` should be fine and these 3 cases can be dropped.

File: server/src/test/java/io/crate/expression/scalar/ConcatwsFunctionTest.java
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/ConcatwsFunctionTest.java:60]: same as above, if we drop special handling of the last null, `testWithNull` should be enough. We can drop this and probably try another interesting case - `all_non_separator_args_are_nulls_returns_empty_string`
checked in https://www.db-fiddle.com for PG13, 
`select concat_ws(',', null)` returns an empty string.

**Review by BaurzhanSakhariev — COMMENTED** [review]
Thanks for your contribution, looks good! Overall on a right track, left some minor comments.

File: docs/appendices/release-notes/unreleased.rst
Comment (rc-enfuse) [reply docs/appendices/release-notes/unreleased.rst]: Thank you for you comment. I've implemented your suggested change.

File: server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java
Comment (rc-enfuse) [reply server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java]: Thank you for you comment. I've implemented your suggested change.

File: server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java
Comment (rc-enfuse) [reply server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java]: Thank you for you comment. This is cleaner. I've implemented your suggested change.

File: server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java:93]: PostgreSQL returns `null` if the first argument is null. I think we should match the behavior:

```
# select concat_ws(null, 'foo', 'bar');
 concat_ws
-----------
 💩
(1 row)
```

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java:93]: "should not be NULL" unfortunately doesn't state how it behaves if it is null. That it fails isn't stated explicitly and the actual behavior is that it returns null (at least in PostgreSQL 13.6, I didn't check with other versions)

Our default policy is to follow PostgreSQL semantics unless there is a good reason to diverge.

**Comment by rc-enfuse** [comment]
All the failed checks seem to have failed via some processes we wouldn't have access too to apply a fix. Please let me know if there is anything I can do to help this pull request along to closure. 
Failing Checks:
**Test CrateDB SQL on windows-latest:** 

> Error: The operation was canceled

.
**Sphinx linkcheck:**

> Run sphinx-build -n -W --keep-going -q -c docs/ -b linkcheck -E docs/ docs/out/html
> /home/runner/work/crate/crate/docs/admin/discovery.rst.rst:128: WARNING: broken link: https://azure.microsoft.com/en-us/documentation/articles/resource-group-authenticate-service-principal-cli/ (404 Client Error: Not Found for url: https://docs.microsoft.com/en-us/api/Redirect/en-us/documentation/articles/resource-group-authenticate-service-principal-cli/)
> INFO: Adjusted `proxied_api_host` to https://crate.io/docs/crate/reference/_
> INFO: No adjustments to `html_context`
> sphinx-sitemap warning: No pages generated for site.xml
> Error: Process completed with exit code 1.

**Vale**

> Vale: Found 0 suggestion(s), 0 warning(s), and 1 error(s).
> Warning: Unable to create annotations; printing Vale alerts instead.
> [ { path: 'docs/appendices/release-notes/unreleased.rst',
>     start_line: [11](https://github.com/crate/crate/runs/6077286406?check_suite_focus=true#step:4:11)1,
>     end_line: 111,
>     start_column: [17](https://github.com/crate/crate/runs/6077286406?check_suite_focus=true#step:4:17),
>     end_column: [18](https://github.com/crate/crate/runs/6077286406?check_suite_focus=true#step:4:18),
>     annotation_level: 'failure',
>     title: '[error] Vale.Spelling',
>     message: 'Did you really mean \'ws\'?' } ]

**Review by BaurzhanSakhariev — COMMENTED** [review]
Thanks for addressing comments! Sorry, I forgot to mention, that doc changes must be max 80 symbols per line - I added suggestions and also added minor comment.

**Comment by BaurzhanSakhariev** [comment]
> All the failed checks seem to have failed via some processes we wouldn't have access too to apply a fix. Please let me know if there is anything I can do to help this pull request along to closure.

Hi @rc-enfuse, thanks for your efforts. I left another round of minor suggestions + there is already a comment about first arg  = null -> return null, otherwise looks good.

Sphinx might be failing because of 80 symbols per line, should be fine after applying suggestions. I will take care of Vale failure as a last step before squashing commits, no worries.

File: docs/general/builtins/scalar-functions.rst
Comment (BaurzhanSakhariev) [reply docs/general/builtins/scalar-functions.rst:62]: 80 per line + ups about null separator - returns null

**Comment by rc-enfuse** [comment]
@BaurzhanSakhariev Thanks. I'll get right on these changes.

File: server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java
Comment (rc-enfuse) [reply server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java:93]: @mfussenegger Ill make the changes. Thanks for the feedback.

File: server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java
Comment (rc-enfuse) [reply server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java]: changed delimiter to separator in tests and code.

**Comment by rc-enfuse** [comment]
Pending approval I think this pull request is ready to merge. unless someone else has any other comments.

**Comment by BaurzhanSakhariev** [comment]
Hi @rc-enfuse, thanks for applying suggestions, looks good. Could you give me permissions to push into your branch? I did some polishing and squashed commits

**Comment by rc-enfuse** [comment]
> Hi @rc-enfuse, thanks for applying suggestions, looks good. Could you give me permissions to push into your branch? I did some minor polishing and squashed commits

sure but doesn't github usually give you the ability to squash commits before we merge?

**Comment by BaurzhanSakhariev** [comment]
> sure but doesn't github usually give you the ability to squash commits before we merge?

I needed permission to do some minor changes as well (didn't want to bother you by new comments as it's really small naming updates/reverting non-related changes)

**Comment by BaurzhanSakhariev** [comment]
retest this please

**Review by BaurzhanSakhariev — APPROVED** [review]
Thanks!

**Comment by BaurzhanSakhariev** [comment]
Hi @rc-enfuse, looks good and ready to merge.

Short update: We are about to release version 4.8 and currently have feature freeze on the master. Once we release we merge this PR.

Thanks for your contribution!

**Pull request merged**
